### PR TITLE
The training step: a fused optimizer, a folded soft Dice

### PR DESCRIPTION
**Pull request by vboussot**
TITLE: The training step: a fused optimizer, a folded soft Dice

BODY:
Three changes to the per-step cost of training. `fused` and `foreach` both unset now ask for the widest batched optimizer step the optimizer implements: fused where torch has one, foreach otherwise, torch's own default off CUDA, and the device read is the one the run will place the graph on, not the one the parameters sit on when the optimizer is built on the launcher. Only a fused optimizer applies the AMP scale itself, so `GradScaler.step` stops reading `found_inf` back to the host and the host enqueues the next forward while the GPU finishes the backward. The soft Dice's per-label loop becomes one expression: the labels' channels gathered in one slice, the reference one comparison against the label vector on an axis of its own, each sum one reduction, dividing by the reference's exact voxel count the integer `_score` already divides by. The metric route keeps its loop, because it carries no gradient and its peak stays one channel, and it now masks its sums in the probability's own dtype instead of expanding the reference to float32. Alongside them, the group's accumulated backward is gated on the criterion's own two flags rather than a `np.unique` over the group's accumulation losses, Dice reads its lowest label in one reduction per map, and the `autocast` and `channels_last` config-guide rows, which carried the shipped 2D example only, gain their 3D numbers.

## Measured

`examples/Segmentation`, autocast + channels_last, one RTX PRO 5000, 325 steps per epoch (`Config_cl.yml`, gpu 0, run in-process), soft Dice folded:

| | before | after |
|---|---|---|
| epoch, median of 9 alternating pairs | 12.95 s (12.73-13.37) | 9.67 s (9.38-10.16) |
| criteria phase | 22.6 ms/step | 14.8 ms/step |
| cudaLaunchKernel (torch.profiler, 3 mid-epoch steps) | 802.0 per step | 307.0 per step |
| CUDA kernels | 1073.0 per step | 468.3 per step |
| `aten::add_` | 63.3 calls, 8.450 ms | 21.0 calls, 1.006 ms |
| `aten::fill_` | 67.7 calls, 2.519 ms | 47.0 calls, 0.240 ms |
| criteria device time | 2.407 ms/step | 1.732 ms/step |

Peak CUDA allocation of the call, `labels=None` over 41 channels: `[8, 41, 256, 256]` 242 -> 180 MiB with gradient, 6 -> 180 MiB under `no_grad`; `[1, 41, 128, 128, 128]` 968 -> 720 MiB with gradient, 18 -> 720 MiB under `no_grad`. Training peaks lower: the backward held a float copy of both operands per label already. A whole soft volume scored in one call with no gradient to build peaks at the label count where the loop peaked at one channel.

Fused optimizer, `konfai TRAIN -c Config_cl.yml --gpu 0 -y` alternated with the same config at `fused: false`. On a quiet host (load 2.3-4.2), three pairs: epoch 13.0 s median (12.9-13.2) fused against 13.7 s (13.5-13.9) foreach, forward + criteria + backward+step 10.1 s against 11.1 s, 31.1 ms a step against 34.2. Twelve more pairs at load 5-29: compute phases 10.1 s median (8.7-10.7) against 11.6 s (11.2-11.8). torch.profiler: 9 `cudaStreamSynchronize` a step against 10; the step's own host time 0.474 ms -> 0.193 ms, its kernels 8 -> 2. Off CUDA nothing is forced; forcing foreach there would cost 526 ms a step against 2.0 ms.

Metric-route masking, five calls each, `labels=None` over 41 channels: `[8, 41, 256, 256]` 1.22 -> 0.98 ms, peak 86.5 -> 85.0 MiB; `[1, 41, 128, 128, 128]` 1.92 -> 1.76 ms, peak 346 -> 340 MiB; `[1, 41, 64, 64, 64]` 1.09 -> 0.97 ms, peak 43.3 -> 42.5 MiB. Dice's lowest label: 0.1040 -> 0.0967 ms for the counts of a `[8, 1, 256, 256]` int64 map (CUDA events). The accumulation gate: 2.94 us per call (timeit, 20000 calls) against 0.02 us.

Docs numbers, 3D UNet of the same shape as the shipped example (five levels to 256 channels, `dim: 3`), 96 cubed patches, batch 2, twenty synthetic 128 cubed cases, three alternating passes each. Training epoch (`konfai TRAIN --gpu 0 -y`), median of nine: fp32 27.0 s (25.5-27.9), autocast 11.7 s (10.9-13.2), autocast + `channels_last` 10.0 s (9.8-10.9). Prediction sweep (`konfai PREDICTION --gpu 0 -y`), the run's own accounting: fp32 4.8 s, autocast 2.9 s, autocast + `channels_last` 2.4 s; whole process 6.8, 4.8 and 4.5 s.

## Values

Two reductions change order on CUDA. The folded soft Dice is not bit-identity: over 5 seeded batches of the example's shape (24 held labels of 41 channels), max |diff| 0 on the loss, 7.5e-9 on a per-label value, 4.5e-13 on the logits gradient, 1.9e-7 of its scale; 0 in fp16, and the same on a 3D `[2, 41, 64, 64, 64]` batch. The fused kernel sums in another order, so after 100 AdamW steps over those 1.93 M parameters the weights sit 3.7e-05 from the foreach ones, 8.3e-06 of their own scale (foreach against single-tensor, same run: 2.4e-07); off CUDA the 100 steps stay bit-identical. Neither reaches the labels: one seeded epoch each way then the 5 cases predicted gives 0 differing voxels of 58 446 992 for the soft Dice, the same count a before/before pair gives, and the two checkpoints of the same seeded epoch predict the 5 cases identically, 0 of 58 446 992 voxels differ, for the optimizer. The accumulation gate, the lowest-label reduction and the metric-route masking are unchanged, the last bit-identical with max |diff| 0 on every sum of the three shapes.

## Tests

`tests/unit/test_network_step.py` pins the off-CUDA bit-identity of the 100 steps and the accumulation contract stays pinned by `test_accumulation_backward_not_refired_by_plain_loss_in_same_group`; `tests/unit/test_measure.py` spells the soft Dice formula out in its oracle, and a reference label the probability map has no channel for is refused with a `MeasureError` instead of raising `IndexError` from the Python slice or hitting a CUDA device-side assert. `flatten` and `dice_per_channel` had no caller left and are gone.

## Stack

Stacked on `pr/1.8.2-12-out-of-core`; merge after it. Before deleting this branch, retarget the next PR of the stack.

## Summary by CodeRabbit

- **New Features**
  - Optimizer execution now automatically selects optimized GPU settings while respecting explicit configuration.
  - Training configuration documentation adds optimizer guidance and performance details.

- **Bug Fixes**
  - Improved gradient accumulation behavior for mixed loss groups.
  - Enhanced Dice metric handling for batched inputs, labels, validation, and empty references.

- **Documentation**
  - Added benchmark results for mixed-precision and channels-last settings in training and inference scenarios.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

File: tests/unit/test_network_step.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+@pytest.fixture
+def on_gpu(monkeypatch: pytest.MonkeyPatch):
+    """The run places the graph on a GPU. The parameters stay on the host: an optimizer is built on
+    the launcher, before ``Network.to`` moves them, which is why the choice reads the run, not them."""
+    monkeypatch.setenv("CUDA_VISIBLE_DEVICES", "0")
+
+
+@pytest.fixture
+def on_cpu(monkeypatch: pytest.MonkeyPatch):
+    monkeypatch.setenv("CUDA_VISIBLE_DEVICES", "")
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_network_step.py:35]: Not needed: `tests/conftest.py::_konfai_env` is autouse and already sets both with `monkeypatch.setenv` for every test.
